src/VisualStudio/Core/Def/Implementation/UnusedReferences/ProjectAssets/ProjectAssetsReader.cs

```diff
@@ -0,0 +1,126 @@
+                .Select(projectReference => BuildReference(projectAssets, target, projectReference, autoReferences))
+                .OfType<ReferenceInfo>()
+                .ToImmutableArray();
+
+            return references;
+        }
+
+        private static ReferenceInfo? BuildReference(
+            ProjectAssetsFile projectAssets,
+            Dictionary<string, ProjectAssetsTargetLibrary> target,
+            ReferenceInfo projectReference,
+            ImmutableHashSet<string> autoReferences)
+        {
+            var referenceName = projectReference.ReferenceType == ReferenceType.Project
+                ? Path.GetFileNameWithoutExtension(projectReference.ItemSpecification)
+                : projectReference.ItemSpecification;
+
+            if (autoReferences.Contains(referenceName))
+            {
+                return null;
+            }
+
+            return BuildReference(projectAssets, target, referenceName, projectReference.TreatAsUsed);
+        }
+
+        private static ReferenceInfo? BuildReference(
+            ProjectAssetsFile projectAssets,
+            Dictionary<string, ProjectAssetsTargetLibrary> target,
+            string dependency,
+            bool treatAsUsed)
+        {
+            var key = target.Keys.FirstOrDefault(library => library.Split('/')[0] == dependency);
+            if (key is null)
+            {
+                return null;
+            }
+
+            return BuildReference(projectAssets, target, dependency, treatAsUsed, key, target[key]);
+        }
+
+        private static ReferenceInfo? BuildReference(
+            ProjectAssetsFile projectAssets,
+            Dictionary<string, ProjectAssetsTargetLibrary> target,
+            string referenceName,
+            bool treatAsUsed,
+            string key,
+            ProjectAssetsTargetLibrary targetLibrary)
+        {
+            if (projectAssets.Libraries is null ||
+                !projectAssets.Libraries.TryGetValue(key, out var library))
+            {
+                return null;
+            }
+
+            var type = targetLibrary.Type switch
+            {
+                "package" => ReferenceType.Package,
+                "project" => ReferenceType.Project,
+                _ => ReferenceType.Assembly
+            };
+
+            var dependencies = targetLibrary.Dependencies != null
+                ? targetLibrary.Dependencies.Keys
+                    .Select(dependency => BuildReference(projectAssets, target, dependency, treatAsUsed: false))
+                    .OfType<ReferenceInfo>()
+                    .ToImmutableArray()
+                : ImmutableArray<ReferenceInfo>.Empty;
+
+            var packagesPath = projectAssets.Project?.Restore?.PackagesPath ?? string.Empty;
+            var compilationAssemblies = targetLibrary.Compile != null
+                ? targetLibrary.Compile.Keys
+                    .Where(assemblyPath => !assemblyPath.EndsWith("_._"))
```

Comment:
What does `_._` mean?

Reply:
Apparently for nuget packages `_._` is used to keep empty folders from being removed. These entries are not real files.

Reply:
I'll add a comment.